Migrate use of SuspendableTimer to EventLoop

rniwa opened the pull request:
#### 4bd78ab3969b623c27406ed7811aee9fa7def839
<pre>
Migrate use of SuspendableTimer to EventLoop
<a href="https://bugs.webkit.org/show_bug.cgi?id=259881">https://bugs.webkit.org/show_bug.cgi?id=259881</a>

Reviewed by Wenson Hsieh.

This PR introduces two new member functions scheduleTask and cancelScheduledTask to
EventLoop and EventLoopTaskGroup to replace the use of SuspendableTimer.

scheduleTask internally creates a new SuspendableTimer and returns the pointer value of
the timer as uintptr_t. This value can then be used to call cancelScheduledTask.

This PR replaces most uses of SuspendableTimer with scheduleTask and cancelScheduledTask.

* Source/WebCore/dom/EventLoop.cpp:
(WebCore::EventLoop::scheduleTask):
(WebCore::EventLoop::cancelScheduledTask):
(WebCore::EventLoopTaskGroup::scheduleTask):
(WebCore::EventLoopTaskGroup::cancelScheduledTask):
* Source/WebCore/dom/EventLoop.h:
* Source/WebCore/editing/AlternativeTextController.cpp:
(WebCore::AlternativeTextController::AlternativeTextController):
(WebCore::AlternativeTextController::startAlternativeTextUITimer):
(WebCore::AlternativeTextController::stopAlternativeTextUITimer):
* Source/WebCore/editing/AlternativeTextController.h:
* Source/WebCore/html/ImageBitmap.cpp:
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::WebGLRenderingContextBase):
(WebCore::WebGLRenderingContextBase::forceRestoreContext):
(WebCore::WebGLRenderingContextBase::scheduleTaskToDispatchContextLostEvent):
(WebCore::WebGLRenderingContextBase::maybeRestoreContextSoon):
(WebCore::WebGLRenderingContextBase::maybeRestoreContext):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
* Source/WebCore/page/EventSource.cpp:
(WebCore::EventSource::EventSource):
(WebCore::EventSource::scheduleInitialConnect):
(WebCore::EventSource::scheduleReconnect):
(WebCore::EventSource::close):
* Source/WebCore/page/EventSource.h:

Canonical link: <a href="https://commits.webkit.org/266664@main">https://commits.webkit.org/266664@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/be2aa043c965b4639d18650b6cc1cdfff5df8d6b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14396 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14706 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15048 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16135 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/13625 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/14529 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17221 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/14782 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/16288 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14575 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15113 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12217 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/16855 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12398 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12978 "1 flakes 5 failures") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/19986 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13475 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13141 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16355 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13697 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/11540 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12984 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17322 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1718 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13541 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/dom/EventLoop.cpp:
```diff
@@ -40,6 +67,27 @@ void EventLoop::queueTask(std::unique_ptr<EventLoopTask>&& task)
 }
 
+// FIXME: Remove the dependency on ScriptExecutionContext.
+uintptr_t EventLoop::scheduleTask(Seconds timeout, ScriptExecutionContext& context, std::unique_ptr<EventLoopTask>&& action)
```
Nit - for slightly improved clarify, maybe we could use a typedef along the lines of `using EventLoopTimerPtr = uintptr_t` and return `EventLoopTimerPtr`.

rniwa replied:
Sure, done.

whsieh commented on Source/WebCore/html/ImageBitmap.cpp:
```diff
@@ -779,9 +778,11 @@ class PendingImageBitmap final : public ActiveDOMObject, public FileReaderLoader
-        ASSERT(!m_createImageBitmapTimer.isActive());
         m_arrayBufferToProcess = WTFMove(arrayBuffer);
-        m_createImageBitmapTimer.startOneShot(0_s);
+        scriptExecutionContext()->eventLoop().queueTask(TaskSource::InternalAsyncTask, [weakThis = WeakPtr { *this }]() {
```
Nit - I think you can omit `()` here.

(And in a few other places in this patch)

rniwa replied:
Will remove.

rniwa commented on Source/WebCore/editing/AlternativeTextController.cpp:
```diff
@@ -114,12 +113,19 @@ void AlternativeTextController::startAlternativeTextUITimer(AlternativeTextType
     m_type = type;
-    m_timer.startOneShot(correctionPanelTimerInterval);
+    stopAlternativeTextUITimer();
+    m_timer = m_document.eventLoop().scheduleTask(correctionPanelTimerInterval, m_document, TaskSource::UserInteraction, [weakThis = WeakPtr { *this }]() {
```
Removing () here.

rniwa commented on Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
```diff
@@ -5676,7 +5673,24 @@ void WebGLRenderingContextBase::scheduleTaskToDispatchContextLostEvent()
+    if (m_restoreTimer)
+        scriptExecutionContext->eventLoop().cancelScheduledTask(m_restoreTimer);
+
+    m_restoreTimer = scriptExecutionContext->eventLoop().scheduleTask(timeout, *scriptExecutionContext, TaskSource::WebGL, [weakThis = WeakPtr { *this }]() {
```
Removing () here.

rniwa commented on Source/WebCore/page/EventSource.cpp:
```diff
@@ -135,14 +134,26 @@ void EventSource::scheduleInitialConnect()
+    auto* context = scriptExecutionContext();
+    if (m_connectTimer)
+        context->eventLoop().cancelScheduledTask(m_connectTimer);
+    m_connectTimer = context->eventLoop().scheduleTask(0_s, *context, TaskSource::DOMManipulation, [ weakThis = WeakPtr { *this } ]() {
```
Removing () here

rniwa commented on Source/WebCore/page/EventSource.cpp:
```diff
@@ -135,14 +134,26 @@ void EventSource::scheduleInitialConnect()
+    auto* context = scriptExecutionContext();
+    if (m_connectTimer)
+        context->eventLoop().cancelScheduledTask(m_connectTimer);
+    m_connectTimer = context->eventLoop().scheduleTask(1_ms * m_reconnectDelay, *context, TaskSource::DOMManipulation, [ weakThis = WeakPtr { *this } ]() {
```
and here

rniwa committed:
Migrate use of SuspendableTimer to EventLoop
https://bugs.webkit.org/show_bug.cgi?id=259881

Reviewed by Wenson Hsieh.

This PR introduces two new member functions scheduleTask and cancelScheduledTask to
EventLoop and EventLoopTaskGroup to replace the use of SuspendableTimer.

scheduleTask internally creates a new SuspendableTimer and returns the pointer value of
the timer as uintptr_t. This value can then be used to call cancelScheduledTask.

This PR replaces most uses of SuspendableTimer with scheduleTask and cancelScheduledTask.

* Source/WebCore/dom/EventLoop.cpp:
(WebCore::EventLoop::scheduleTask):
(WebCore::EventLoop::cancelScheduledTask):
(WebCore::EventLoopTaskGroup::scheduleTask):
(WebCore::EventLoopTaskGroup::cancelScheduledTask):
* Source/WebCore/dom/EventLoop.h:
* Source/WebCore/editing/AlternativeTextController.cpp:
(WebCore::AlternativeTextController::AlternativeTextController):
(WebCore::AlternativeTextController::startAlternativeTextUITimer):
(WebCore::AlternativeTextController::stopAlternativeTextUITimer):
* Source/WebCore/editing/AlternativeTextController.h:
* Source/WebCore/html/ImageBitmap.cpp:
* Source/WebCore/html/canvas/WebGLRenderingContextBase.cpp:
(WebCore::WebGLRenderingContextBase::WebGLRenderingContextBase):
(WebCore::WebGLRenderingContextBase::forceRestoreContext):
(WebCore::WebGLRenderingContextBase::scheduleTaskToDispatchContextLostEvent):
(WebCore::WebGLRenderingContextBase::maybeRestoreContextSoon):
(WebCore::WebGLRenderingContextBase::maybeRestoreContext):
* Source/WebCore/html/canvas/WebGLRenderingContextBase.h:
* Source/WebCore/page/EventSource.cpp:
(WebCore::EventSource::EventSource):
(WebCore::EventSource::scheduleInitialConnect):
(WebCore::EventSource::scheduleReconnect):
(WebCore::EventSource::close):
* Source/WebCore/page/EventSource.h:

Canonical link: https://commits.webkit.org/266664@main
